File: sdk/typescript/examples/cost_tracking.ts
/**
 * Cost Tracking — Monitor spending and savings in real-time.
 *
 * Uses the telemetry callback to log every request's cost and savings.
 *
 * Prerequisites:
 *   export AI_INFRA_API_KEY="sk-your-key-here"
 *   npx ts-node examples/cost_tracking.ts
 */

import { Client, type RequestMetrics } from "../src";

function onRequest(metrics: RequestMetrics): void {
  const tag = metrics.isCacheHit ? "[CACHE]" : "[INFER]";
  console.log(
    `${tag} ${metrics.model} | ` +
      `${metrics.latencyMs}ms | ` +
      `$${metrics.costUsd.toFixed(4)} cost | ` +
      `$${metrics.savingsUsd.toFixed(4)} saved`,
  );
}

async function main() {
  const client = new Client({
    telemetry: true,
    onRequest, // Called after every request
  });

  // Send multiple requests to accumulate stats
  const prompts = [
    "What is 2+2?",
    "Explain DNS in one sentence.",
    "What is 2+2?", // Likely cache hit
    "Translate 'hello' to French.",
  ];

  for (const prompt of prompts) {
    const response = await client.chat.completions.create({
      messages: [{ role: "user", content: prompt }],
    });
    console.log(`  → ${response.choices[0].message.content.slice(0, 80)}`);
    console.log();
  }

  // Final stats
  const stats = client.getStats();
  console.log("═══════════════════════════════════════");
  console.log("         Cost Tracking Summary");
  console.log("═══════════════════════════════════════");
  console.log(`Total requests:  ${stats.totalRequests}`);
  console.log(`Total cost:      $${stats.totalCostUsd.toFixed(4)}`);
  console.log(`Total saved:     $${stats.totalSavingsUsd.toFixed(4)}`);
  console.log(`Cache hit rate:  ${(stats.cacheHitRate * 100).toFixed(1)}%`);
  console.log(`Avg latency:     ${stats.avgLatencyMs.toFixed(0)}ms`);
  console.log(`Error rate:      ${(stats.errorRate * 100).toFixed(1)}%`);
}

main().catch(console.error);

File: sdk/python/examples/cost_tracking.py
"""Cost tracking — see real-time savings vs OpenAI pricing.

Run:
    export AI_INFRA_API_KEY="sk-your-key-here"
    python examples/cost_tracking.py
"""

from ai_infra import Client

# Enable telemetry to track aggregate stats
client = Client(telemetry=True, mode="best_cost")

prompts = [
    "What is machine learning?",
    "Explain neural networks briefly.",
    "What are transformers in AI?",
    "How does gradient descent work?",
    "What is transfer learning?",
]

print("Sending 5 requests with best_cost routing...\n")

total_saved = 0.0
total_cost = 0.0

for prompt in prompts:
    response = client.chat.completions.create(
        model="auto",
        messages=[{"role": "user", "content": prompt}],
        max_tokens=256,
    )

    savings = response.savings
    total_cost += savings.cost_usd
    total_saved += savings.cost_saved_usd

    print(f"  {prompt[:40]:<40} | "
          f"model={savings.model_used:<15} | "
          f"${savings.cost_usd:.4f} | "
          f"saved ${savings.cost_saved_usd:.4f}")

# Print aggregate stats
print(f"\n{'='*70}")
print(f"Total cost:    ${total_cost:.4f}")
print(f"Total saved:   ${total_saved:.4f}")
if total_cost + total_saved > 0:
    pct = total_saved / (total_cost + total_saved) * 100
    print(f"Savings rate:  {pct:.1f}%")

# Show telemetry stats
stats = client.get_stats()
print("\nTelemetry stats:")
print(f"  Requests:      {stats['total_requests']:.0f}")
print(f"  Avg latency:   {stats['avg_latency_ms']:.0f}ms")
print(f"  Cache hit rate: {stats['cache_hit_rate']:.0%}")
print(f"  Error rate:    {stats['error_rate']:.0%}")

client.close()
